### DISCUSSION TITLE
ObservableObject@2.init? being read outside a reactive context

### DISCUSSION THREAD

**Discussion by codeBelt** [mobxjs/mobx#2722] [Q&A] [answered]
Can someone hep me understand why I get the following warning and what I should be doing differently.

`[mobx] Observable ObservableObject@2.init? being read outside a reactive context `

**Code**  https://codesandbox.io/s/observable-init-read-outside-1bjvz?file=/src/App.tsx

```ts
export const App = observer(() => {
  const [localStore] = useState(() =>
    observable({
      name: '',

      init(name) {
        this.name = name;
      },
    })
  );

  useEffect(() => {
    localStore.init('CodeSandbox');
  }, [localStore]);

  return (
    ...
  );
});
```

**Comment by mweststrate** [comment]
I think this should fix it: pass `{init: action}` as second argument to
`observable` (or use `useLocalObservable` hook from mobx-react-lite)

**Reply by codeBelt** [reply]
Thanks for the fast response. I tried what you suggested and still get the warning for all of the following: 
https://codesandbox.io/s/observable-init-read-outside-example-2-hprn1?file=/src/App.tsx

```ts
import { action, configure, observable } from "mobx";
import { observer, useLocalObservable } from "mobx-react-lite";

const store = {
  init() {} // action
};
...
const localStore = useLocalObservable(() => store);
...
const localStore = useLocalObservable(() => store, { init: action });
...
const [localStore] = useState(() => observable(store));
...
const [localStore] = useState(() => observable(store, { init: action }));

useEffect(() => {
  localStore.init();
}, [localStore]);
...
```

Note: ES6 Classes with `makeObservable` or `makeAutoObservable` work fine without any warnings.

**Reply by codeBelt** [reply]
@mweststrate is this an issue with factory functions? Should I create a bug ticket?

**Comment by mweststrate** [comment]
Please be patient, we're all just volunteering here.

**Comment by codeBelt** [comment, marked as answer]
Note sure why I didn't try `makeAutoObservable` with the factory function and only with a Class but doing so made the warning go away.

**Code:** https://codesandbox.io/s/makeautoobservable-init-read-outside-forked-2yirn?file=/src/App.tsx

```js
export const App = observer(() => {
  const [localStore] = useState(() =>
    makeAutoObservable({
      name: "",

      init(name: string) {
        this.name = name;
      }
    })
  );

  useEffect(() => {
    localStore.init("CodeSandbox");
  }, [localStore]);

  return (
    ...
  );
});
```

I will change my code to use `makeAutoObservable`.

**Reply by urugator** [reply]
I think this is a "bug" in proxy impl that creates `pendingKey` even outside derivation:
https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/types/dynamicobject.ts#L50
Hopefully shouldn't occur in https://github.com/mobxjs/mobx/pull/2641.
However, we are considering that functions, that are not explicitely annotated, will be observable as well - in such case it would throw anyway and you would either have to disable the `observableRequiresReaction` or you would have to:
```javascript
useEffect(() => {
    runInAction(() => {
       localStore.init("CodeSandbox");
    })    
  }, [localStore]);
```
